polkadot/node/core/pvf/src/execute/queue.rs

```diff
@@ -278,15 +292,59 @@ impl Queue {
 		}
 
 		let job = queue.remove(job_index).expect("Job is just checked to be in queue; qed");
+		let exec_kind = job.exec_kind;
+
+		if !self.is_job_relevant(exec_kind) {
+			let _ = job.result_tx.send(Err(ValidationError::ExecutionDeadline));
+			gum::warn!(
+				target: LOG_TARGET,
+				?priority,
+				?exec_kind,
+				"Job exceeded its deadline and was dropped without execution",
+			);
+			return self.try_assign_next_job(finished_worker);
+		}
 
 		if let Some(worker) = worker {
 			assign(self, worker, job);
 		} else {
 			spawn_extra_worker(self, job);
 		}
-		self.metrics.on_execute_kind(priority);
+		self.metrics.on_execute_kind(exec_kind);
 		self.unscheduled.mark_scheduled(priority);
 	}
+
+	fn update_active_leaves(&mut self, update: ActiveLeavesUpdate, ancestors: Option<Vec<Hash>>) {
+		for hash in update.deactivated {
+			let _ = self.active_leaves.remove(&hash);
+		}
+
+		let Some(leaf) = update.activated else { return };
+		let ancestors = ancestors.unwrap_or_default();
+		let _ = self.active_leaves.insert(
+			leaf.hash,
+			ActiveLeafWithAncestors { hash: leaf.hash, number: leaf.number, ancestors },
+		);
+	}
+
+	fn is_job_relevant(&self, exec_kind: PvfExecKind) -> bool {
+		let Some(ttl) = exec_kind.ttl() else { return true };
+		let Some(current_block_number) = self
+			.active_leaves
+			.values()
+			.filter(|x| x.hash == ttl.relay_parent || x.ancestors.contains(&ttl.relay_parent))
+			.map(|x| x.number)
+			.min()
+			.or_else(|| {
+				gum::debug!(target: LOG_TARGET, ?ttl, "Relay parent of the execution job was not found among the ancestors of active leaves");
+				self.active_leaves.values().map(|x| x.number).min() // Use the earliest block number as the current one.
```

Comment:
I think we could just drop the candidate here. It arrived too late or can it be possible that we got the message before the active leaves update. Subsystem prioritizes the processing of signals (active leaves) so the PVF backend should not be able to receive a validation request before the active leaves update.